### PR TITLE
Mark sync as limited if there is a gap in the timeline 

### LINKED ISSUE #16463: `/sync` doesn't inform clients when there is a gap in the timeline
If there has bee e.g. a netsplit in a room where the remote servers have sent more than ten events, then when the netsplit resolves the local server will fetch up to 10 events. The remaining events will be backfilled when a client does an appropriate call to `/messages`.

The syncing clients will receive the new events down their timeline, however the `limited` flag is not set. This means the client thinks there is no gap, and so won't try and backfill any missed messages. 

The net result is that clients won't ever see the older events in the remote fork.

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #16463

### CODE REVIEW

File: synapse/storage/databases/main/events.py [hunk 1]
```diff
@@ -2267,35 +2267,40 @@ def _update_backward_extremeties(
...
 
-        txn.execute_batch(
-            query,
-            [
-                (e_id, ev.room_id, e_id, ev.room_id, e_id, ev.room_id)
-                for ev in events
-                for e_id in ev.prev_event_ids()
-                if not ev.internal_metadata.is_outlier()
-            ],
+        room_id = events[0].room_id
```
Comment (clokep) [inline synapse/storage/databases/main/events.py:2271]: I guess something has gone horribly wrong if these don't all have the same room ID?

File: synapse/storage/databases/main/stream.py [hunk 2]
```diff
@@ -1616,3 +1616,48 @@ async def get_name_from_instance_id(self, instance_id: int) -> str:
...
             desc="get_name_from_instance_id",
         )
+
+    async def get_timeline_gaps(
+        self,
+        room_id: str,
+        from_token: Optional[RoomStreamToken],
+        to_token: RoomStreamToken,
+    ) -> Optional[RoomStreamToken]:
+        """Check if there is a gap, and return the token."""
```
Comment (clokep) [inline synapse/storage/databases/main/stream.py:1626]: "the token" sounds like it can be expanded a bit to say whether this is the start of the gap, etc. (see the implementation comments at the end for calculating the `last_position`)

File: synapse/storage/databases/main/events.py [hunk 3]
```diff
@@ -2267,35 +2267,59 @@ def _update_backward_extremeties(
...
+
+            # Record the stream orderings where we have new gaps.
+            gap_events = [
+                (room_id, self._instance_name, ev.internal_metadata.stream_ordering)
+                for ev in events
+                if any(
+                    e_id in potential_backwards_extremities
+                    for e_id in ev.prev_event_ids()
+                )
+            ]
```
Comment (clokep) [inline synapse/storage/databases/main/events.py:2314]: Do gaps ever get "filled"? I think they *don't* and that doesn't make sense, but want to double check.

File: synapse/handlers/sync.py [hunk 4]
```diff
@@ -684,7 +699,9 @@ async def _load_filtered_recents(
         return TimelineBatch(
             events=recents,
             prev_batch=prev_batch_token,
-            limited=limited or newly_joined_room,
+            # Also mark as limited if this is a new room or there has been a gap
+            # (to force client to paginate the gap).
+            limited=limited or newly_joined_room or gap_token is not None,
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/sync.py:704]: Ah, no. We set `limited` back to `False` during the loop to fetch more events if we run out of events to return. For the gap case this happens when as we set the `since_token` to be the `gap_token`, so we have to specifically mark this as always gappy.

File: synapse/storage/databases/main/events.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/events.py:2271]: Yeah, exactly. We should really refactor it all to take a room ID arg to make this explicit

File: synapse/handlers/sync.py [hunk 5]
```diff
@@ -599,7 +610,11 @@ async def _load_filtered_recents(
             end_key = room_key
 
             since_key = None
-            if since_token and not newly_joined_room:
+            if since_token and gap_token:
+                # If there is a gap then we need to only include events after
+                # it.
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/sync.py:615]: If we have a section of the timeline with events `A, B, <gap>, X, Y, Z` we need to make sure we only send down `X, Y, Z` and *not* `A` and `B` (which we currently do. If we did, then the client wouldn't try and paginate to fill in the gap.

File: synapse/handlers/sync.py [hunk 6]
```diff
@@ -543,6 +543,17 @@ async def _load_filtered_recents(
...
+            # Check if there is a gap, if so we need to mark this as limited and
+            # recalculate which events to send down.
+            gap_token = await self.store.get_timeline_gaps(
+                room_id,
+                since_token.room_key if since_token else None,
+                now_token.room_key,
+            )
+            if gap_token:
+                potential_recents = []
+                limited = True
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/sync.py:555]: Yup, as we also need to make sure we don't send down events from before the gap.

File: synapse/storage/databases/main/stream.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/stream.py:1626]:
```suggestion
        """Check if there is a gap, and return a token that marks the position
        of the gap in the stream.
        """
```

Like that?

File: synapse/storage/databases/main/events.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/events.py:2314]: We can backfill and get the events in the gap and "fill" it like that. However, I think there is a lot of subtlety to then removing the entries in `timeline_gaps` table, and not really worth it. (It would only help clients that hadn't synced between when we detected the gap and when we backfilled the events).

File: synapse/storage/databases/main/stream.py (hunk 2 above)
Comment (clokep) [reply synapse/storage/databases/main/stream.py:1626]: That's better. Thanks!

File: synapse/handlers/sync.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/handlers/sync.py:704]: I must have missed the usage in the loop! Thanks!

**Review by clokep — APPROVED** [review]
This should probably have a complement test?

**Comment by erikjohnston** [comment]
> This should probably have a complement test?

Can someone write a bot to remind me about tests being a thing??

**Comment by erikjohnston** [comment]
@clokep I wrote a test and it found an edge case! I've hopefully fixed it....

**Review by clokep — APPROVED** [review]
